pkg/ruler/scheduler.go

```diff
@@ -267,13 +276,13 @@ func (s *scheduler) nextWorkItem() *workItem {
 // workItemDone marks the given item as being ready to be rescheduled.
 func (s *scheduler) workItemDone(i workItem) {
 	s.Lock()
-	ruleSet, found := s.cfgs[i.userID]
+	config, found := s.cfgs[i.userID]
 	var currentRules []rules.Rule
 	if found {
-		currentRules = ruleSet[i.groupName]
+		currentRules = config.rules[i.groupName]
 	}
 	s.Unlock()
-	if !found || len(currentRules) == 0 {
+	if !found || len(currentRules) == 0 || i.generation < config.generation {
```

Comment:
If I understood correctly this condition may be incorrect if `generation` overflows.
Can this happen at all?

If so, we probably want to change this logic to make it more robust.
If not, we probably want a comment about the assumptions made and the reasons why these assumptions hold.

----

Trying to answer my own question, [it seems](https://github.com/weaveworks/cortex/blob/master/cmd/configs/migrations/002_immutable_configs.up.sql#L11) `generation` is backed by a [PostgreSQL `serial`](https://www.postgresql.org/docs/9.5/static/datatype-numeric.html):
```
serial      4 bytes         autoincrementing integer          1 to 2147483647
```
so hopefully this doesn't behave like the `int` used in our code ([which would overflow](https://play.golang.org/p/WNEs3Y-Gabk)), and guarantees that we'd get something like the below before that happens:
```
ERROR:  nextval: reached maximum value of sequence 
```
Next question:
- Can we reach this case?
- What triggers new configs to be added?

----
In `dev`/`local`, the above could still overflow, see [here](https://github.com/weaveworks/cortex/blob/master/pkg/configs/db/memory/memory.go#L40) and [here](https://github.com/weaveworks/cortex/blob/master/pkg/configs/db/memory/memory.go#L71). Not sure if that can be an issue in practice though.

Reply:
The latest config ID increments every time any user does a CRUD operation to the cortex configs service.
The overflow would probably also affect the config loading check - we send latestConfigVersion over to the configs service.

I'm not sure if it makes sense to fix this here? Would it be okay to add a comment and file an issue?

Reply:
I should add: I don't think we've overflowed yet, and are probably not in danger of doing it soon.

Reply:
> I'm not sure if it makes sense to fix this here? Would it be okay to add a comment and file an issue?

Yes, absolutely! :+1: 

> _we probably want a comment about the assumptions made and the reasons why these assumptions hold._

----

> I don't think we've overflowed yet, and are probably not in danger of doing it soon.

Indeed!